### PR TITLE
Expose new /metrics endpoint to allow monitoring of addon-resizer

### PR DESCRIPTION
**Pull request by raywainman**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds Prometheus metrics to addon-resizer, allowing monitoring of overall health. Initial implementation adds a single metric called `addon_resizer_execution_outcome` which is a simple counter metric that tracks the outcome of every addon-resizer loop (for example, whether addon-resizer actually updated the deployment for metrics-server).

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

Follow up to https://github.com/kubernetes/autoscaler/pull/6764 where addon-resizer is made compatible to run in the control plane.

#### Does this PR introduce a user-facing change?

```release-note
-Expose new `/metrics` endpoint in addon-resizer to expose health metrics that are useful for monitoring purposes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

To use:
- Ensure the `expose-metrics` flag is set to `true`.
- The `/metrics` endpoint is available by default on port `:8945` (configurable by the `address` flag).

### CODE REVIEW

**Comment by chelseychen** [comment]
LGTM, thanks for the change!

File: addon-resizer/nanny/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		}, []string{"outcome"},
+	)
+)
+
+// Initialize sets up Prometheus to expose metrics.
+func Initialize(address string) {
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		err := http.ListenAndServe(address, nil)
+		klog.Fatalf("Failed to start metrics: %v", err)
```
Comment (kwiesmueller) [inline addon-resizer/nanny/metrics/metrics.go:49]:
```suggestion
		klog.Fatalf("Failed to serve metrics: %v", err)
```
Nit: But as this also happens when for some reason the server stops later serve might be more accurate in those cases.

File: addon-resizer/nanny/main/pod_nanny.go [hunk 2]
```diff
@@ -84,6 +88,12 @@ func main() {
 	klog.Infof("Version: %s", nanny.AddonResizerVersion)
 	flag.Parse()
 
+	if *exposeMetrics {
+		// Register a handler on the provided address.
+		metrics.Initialize(*address)
+	}
+	metrics.Register()
```
Comment (kwiesmueller) [inline addon-resizer/nanny/main/pod_nanny.go:95]: Is the order of Initialize and Register relevant? I thought register comes first (does Prometheus do registration at runtime?).
Might be worth having a test to see if the metric is served.

**Review by kwiesmueller — CHANGES_REQUESTED** [review]

**Comment by kwiesmueller** [comment]
/lgtm
/hold
In case you want to respond to the comments. I don't think they are blocking (even though a test would be nice, but addon-manager doesn't have too many). Feel free to remove the hold if ack.

**Review by kwiesmueller — APPROVED** [review]
/lgtm
/approve

**Pull request merged**